base_layer/core/src/base_node/state_machine_service/states/sync_decide.rs

```diff
@@ -0,0 +1,130 @@
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    base_node::{
+        state_machine_service::{
+            states::{HeaderSync, StateEvent},
+            BaseNodeStateMachine,
+        },
+        sync::SyncPeer,
+    },
+    chain_storage::BlockchainBackend,
+};
+use log::*;
+
+const LOG_TARGET: &str = "c::bn::state_machine_service::states::sync_decide";
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct DecideNextSync {
+    sync_peers: Vec<SyncPeer>,
+}
+
+impl DecideNextSync {
+    pub async fn next_event<B: BlockchainBackend + 'static>(&mut self, shared: &BaseNodeStateMachine<B>) -> StateEvent {
+        use StateEvent::*;
+        let local_metadata = match shared.db.get_chain_metadata().await {
+            Ok(m) => m,
+            Err(e) => {
+                return FatalError(format!("Could not get local blockchain metadata. {}", e));
+            },
+        };
+
+        debug!(
+            target: LOG_TARGET,
+            "Selecting a suitable sync peer from {} peer(s)",
+            self.sync_peers.len()
+        );
+
+        if shared.config.pruning_horizon > 0 {
+            // Filter sync peers that claim to be able to provide full blocks up until our pruned height
+            let sync_peers_iter = self.sync_peers.iter().filter(|sync_peer| {
+                let chain_metadata = sync_peer.claimed_chain_metadata();
+                let our_pruned_height_from_peer =
+                    local_metadata.horizon_block(chain_metadata.height_of_longest_chain());
+                let their_pruned_height = chain_metadata.pruned_height();
+                our_pruned_height_from_peer >= their_pruned_height
+            });
+
+            match find_best_latency(sync_peers_iter) {
+                Some(sync_peer) => ProceedToHorizonSync(sync_peer),
+                None => Continue,
+            }
+        } else {
+            // Filter sync peers that are able to provide full blocks from our current tip
+            let sync_peers_iter = self.sync_peers.iter().filter(|sync_peer| {
+                sync_peer.claimed_chain_metadata().pruning_horizon() <= local_metadata.height_of_longest_chain()
+            });
+
+            match find_best_latency(sync_peers_iter) {
+                Some(sync_peer) => ProceedToBlockSync(sync_peer),
+                None => Continue,
+            }
+        }
+    }
+}
+
+/// Find the peer with the best latency
+fn find_best_latency<'a, I: IntoIterator<Item = &'a SyncPeer>>(iter: I) -> Option<SyncPeer> {
+    iter.into_iter()
+        .fold(Option::<&'a SyncPeer>::None, |current, sync_peer| match current {
+            Some(p) => match (p.latency(), sync_peer.latency()) {
+                (Some(_), None) => Some(p),
+                (None, Some(_)) => Some(sync_peer),
+                (Some(current), Some(l)) if current > l => Some(sync_peer),
```

Comment:
is this correct?

Reply:
Maybe can name those better, "if current peer's latency is worse (more) than this peer's latency, use this peer"

Reply:
was just wondering about the `1` value, I expected `current > sync_peer` or is that a lower case L ?

Reply:
lol, yeah L for latency, I'll change it - also see there is a test failing - will fix